core/src/main/java/org/apache/iceberg/ReachableFileCleanup.java

```diff
@@ -0,0 +1,178 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * File cleanup strategy for snapshot expiration which determines, via an in-memory reference set,
+ * metadata and data files that are not reachable given the previous and current table states.
+ */
+class ReachableFileCleanup extends FileCleanupStrategy {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReachableFileCleanup.class);
+
+  private final TableMetadata afterExpiration;
+  private final TableMetadata beforeExpiration;
+
+  ReachableFileCleanup(
+      TableOperations ops,
+      TableMetadata beforeExpiration,
+      ExecutorService deleteExecutorService,
+      ExecutorService planExecutorService,
+      Consumer<String> deleteFunc) {
+    super(ops, deleteExecutorService, planExecutorService, deleteFunc);
+    this.beforeExpiration = beforeExpiration;
+    this.afterExpiration = ops.refresh();
+  }
+
+  @Override
+  public void cleanFiles() {
+    // Identify all of the manifest lists to retain
+    Set<String> manifestListsBeforeExpiration =
+        ReachableFileUtil.manifestListLocations(beforeExpiration);
+
+    // Identify all of the manifest lists that currently exist
+    Set<String> manifestListsAfterExpiration =
+        ReachableFileUtil.manifestListLocations(afterExpiration);
+
+    Set<String> manifestListsToDelete =
+        Sets.difference(manifestListsBeforeExpiration, manifestListsAfterExpiration);
+    Map<String, ManifestFile> reachableManifestsAfterExpiration =
+        computeManifestsForManifestLists(manifestListsAfterExpiration);
+
+    // Compute the manifests to delete based on the current manifest lists to delete and the current
+    // reachable manifests
+    List<ManifestFile> manifestsToDelete =
+        computeManifestsToDelete(manifestListsToDelete, reachableManifestsAfterExpiration);
+
+    List<ManifestFile> currentManifests =
+        Lists.newArrayList(reachableManifestsAfterExpiration.values());
+
+    Set<String> filesToDelete = findFilesToDelete(manifestsToDelete, currentManifests);
+    Tasks.foreach(filesToDelete)
+        .executeWith(deleteExecutorService)
+        .retry(3)
+        .stopRetryOn(NotFoundException.class)
+        .suppressFailureWhenFinished()
+        .onFailure((file, exc) -> LOG.warn("Delete failed for data file: {}", file, exc))
+        .run(deleteFunc::accept);
+
+    Set<String> manifestPathsToDelete =
+        manifestsToDelete.stream().map(ManifestFile::path).collect(Collectors.toSet());
+
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+    deleteMetadataFiles(
+        manifestPathsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest", path, exc));
+    deleteMetadataFiles(
+        manifestListsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest list", path, exc));
```

Comment:
@singhpk234 I added the logging in the implementations because having the logger in the abstract class doesn't seem right. It's a bit more duplicate now, but I think it's more explicit and we can keep the logger static . let me know what you think!